#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import json
import time
import re
import serial
import requests
import os
import random
from dataclasses import dataclass, asdict
from typing import Optional, Final
from threading import Thread, Lock

from PyQt6.QtWidgets import *
from PyQt6.QtGui import *
from PyQt6 import uic
from PyQt6.QtCore import Qt, QThread, pyqtSignal, QObject, QTimer

# =========================
# Configuration
# =========================
# [UPDATE] ì‚¬ìš©ì í”¼ë“œë°± ë°˜ì˜: ì„±ëŠ¥ì´ ë” ì¢‹ì€ gemma3:4b ëª¨ë¸ ì‚¬ìš©
OLLAMA_HOST: Final = "http://127.0.0.1:11434"
OLLAMA_URL: Final = f"{OLLAMA_HOST}/api/chat"
MODEL_NAME: Final = "gemma3:4b"  
SERIAL_PORT: Final = "/dev/ttyACM0" 
BAUD_RATE: Final = 115200
PROMPT_FILE: Final = "system_prompt.txt"

ANGER_KEYWORDS = ['ë¹¡ì³', 'ì§œì¦', 'ì„±ì§ˆ', 'êº¼ì ¸', 'ë‹¥ì³', 'ì”¨ë°œ', 'ê°œë¹¡', 'í™”ë‚˜', 'ì£½ì„']

BUTTON_MESSAGES = [
    "ì•ˆë…•?", "ì™€ì¹!", "ì˜¤ëŠ˜ ê¸°ë¶„ ì–´ë•Œ?", 
    "ì‹¬ì‹¬í•´ ë†€ì•„ì¤˜", "ë„ˆëŠ” ëˆ„êµ¬ì•¼?", "ì‚¬ë‘í•´", 
    "ë…¸ë˜ í•œ ì†Œì ˆ ë¶ˆëŸ¬ì¤˜", "ë¬´ì„œìš´ ì´ì•¼ê¸° í•´ì¤˜", 
    "ë…ë„ëŠ” ëˆ„êµ¬ ë•…?", "2 ë”í•˜ê¸° 2ëŠ”?"
]

# ASCII Emoticons for Random Injection
EMOTICONS = {
    "happy": ["^_^", "^o^", "<3", ":)", "B-)", ":D", "XD"],
    "sad": ["T_T", "(ToT)", ";_;", "T.T", "..", ">_<"],
    "angry": ["-^-", ">_<", "-_-", "!!!!", "Orz"],
    "neutral": ["OoO", "OwO", "Hmm", ":]", "?_?", "(?)"]
}

# =========================
# Data Structures
# =========================
@dataclass
class SensorData:
    temp: float = 0.0
    humid: float = 0.0
    light: int = 0
    btn: int = 0
    timestamp: float = 0.0
    
    @property
    def discomfort_index(self) -> float:
        try:
            rh_decimal = self.humid / 100.0
            di = (1.8 * self.temp) - (0.55 * (1 - rh_decimal) * (1.8 * self.temp - 26)) + 32
            return round(di, 1)
        except Exception:
            return 0.0

    @property
    def weather_status(self) -> str:
        di = self.discomfort_index
        if di < 68: return "Pleasant"
        elif 68 <= di < 75: return "Moderate"
        elif 75 <= di < 80: return "Uncomfortable"
        else: return "Dangerous"

    @property
    def light_status(self) -> str:
        if self.light >= 600: return "Dark"
        elif self.light <= 150: return "Bright"
        else: return "Normal"

@dataclass(frozen=True)
class RobotCommand:
    seq: int
    l1: str
    l2: str
    chat_response: str
    mood: str 
    act: str 

    def to_json_serial(self) -> str:
        data = {
            "seq": self.seq,
            "l1": self.l1,
            "l2": self.l2,
            "mood": self.mood,
            "act": self.act
        }
        return json.dumps(data) + "\n"

# =========================
# Hardware Controller
# =========================
class SmartController(QObject):
    sensor_received = pyqtSignal(SensorData)

    def __init__(self, port: str, baud: int):
        super().__init__()
        self.serial: Optional[serial.Serial] = None
        self.port = port
        self.baud = baud
        self._connect_serial()
        
        self._seq_counter: int = 0
        self.is_running: bool = True
        self.lock = Lock()
        self.current_sensor = SensorData()
        
        if self.serial:
            self.receiver_thread = Thread(target=self._listen_serial, daemon=True)
            self.receiver_thread.start()

    def _connect_serial(self):
        try:
            self.serial = serial.Serial(self.port, self.baud, timeout=0.1)
            print(f"âœ… ì‹œë¦¬ì–¼ í¬íŠ¸ ì—°ê²° ì„±ê³µ: {self.port}")
            time.sleep(2) 
        except Exception as e:
            print(f"âš ï¸ ì‹œë¦¬ì–¼ ì—°ê²° ì‹¤íŒ¨ (ì‹œë®¬ë ˆì´ì…˜): {e}")

    def _listen_serial(self) -> None:
        while self.is_running and self.serial and self.serial.is_open:
            try:
                if self.serial.in_waiting > 0:
                    line = self.serial.readline().decode('utf-8', errors='ignore').strip()
                    if not line: continue
                    try:
                        data = json.loads(line)
                        if data.get("type") == "SENSOR":
                            with self.lock:
                                self.current_sensor = SensorData(
                                    temp=float(data.get("temp", 0)),
                                    humid=float(data.get("humid", 0)),
                                    light=int(data.get("light", 0)),
                                    btn=int(data.get("btn", 0)),
                                    timestamp=time.time()
                                )
                            self.sensor_received.emit(self.current_sensor)
                    except json.JSONDecodeError:
                        pass 
            except Exception as e:
                time.sleep(1)

    def send_command(self, cmd: RobotCommand) -> None:
        if self.serial and self.serial.is_open:
            try:
                payload = cmd.to_json_serial().encode('utf-8')
                self.serial.write(payload)
            except Exception as e:
                print(f"[Serial Error] ì „ì†¡ ì‹¤íŒ¨: {e}")

    def get_next_seq(self) -> int:
        self._seq_counter += 1
        return self._seq_counter
        
    def get_sensor_data(self) -> SensorData:
        with self.lock:
            return self.current_sensor

    def close(self) -> None:
        self.is_running = False
        if self.serial and self.serial.is_open:
            self.serial.close()

# =========================
# LLM Worker
# =========================
class LLMWorker(QThread):
    result_signal = pyqtSignal(RobotCommand)
    error_signal = pyqtSignal(str)

    def __init__(self, query: str, seq: int, sensor_data: SensorData):
        super().__init__()
        self.query = query
        self.seq = seq
        self.sensor_data = sensor_data
        
    def _construct_system_prompt(self) -> str:
        base_prompt = ""
        try:
            base_path = os.path.dirname(os.path.abspath(__file__))
            file_path = os.path.join(base_path, PROMPT_FILE)
            if os.path.exists(file_path):
                with open(file_path, 'r', encoding='utf-8') as f:
                    base_prompt = f.read()
            else:
                base_prompt = "You are Artie. JSON Output required."
        except Exception:
            base_prompt = "You are Artie. JSON Output required."
            
        prompt = base_prompt.replace("{temp}", str(self.sensor_data.temp))
        prompt = prompt.replace("{humid}", str(self.sensor_data.humid))
        prompt = prompt.replace("{light}", str(self.sensor_data.light))
        prompt = prompt.replace("{light_status}", self.sensor_data.light_status)
        
        # [UPDATE] ì‹œìŠ¤í…œ ì•Œë¦¼ì„ ì¡°ê¸ˆ ë” ë¶€ë“œëŸ½ê²Œ ë³€ê²½
        if self.sensor_data.light_status == "Dark":
            prompt += "\n[ENV INFO] Dark environment. Act sleepy or scared."
        elif self.sensor_data.light_status == "Bright":
            prompt += "\n[ENV INFO] Bright environment. Act energetic."
            
        return prompt

    def run(self):
        is_furious = any(k in self.query for k in ANGER_KEYWORDS)
        
        system_prompt = self._construct_system_prompt()
        
        payload = {
            "model": MODEL_NAME,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": self.query}
            ],
            "stream": False,
            "format": "json",
            "options": {
                "temperature": 0.6, # ì°½ì˜ì„± ì•½ê°„ ì¦ê°€
                "top_p": 0.9
            }
        }

        try:
            response = requests.post(OLLAMA_URL, json=payload, timeout=25)
            response.raise_for_status()
            
            data = response.json()
            content = data["message"]["content"]
            
            parsed = {}
            try:
                # JSON ì¶”ì¶œ ë¡œì§ (Markdown ì½”ë“œë¸”ë¡ ëŒ€ì‘)
                start_idx = content.find('{')
                end_idx = content.rfind('}')
                
                if start_idx != -1 and end_idx != -1:
                    json_str = content[start_idx : end_idx + 1]
                    parsed = json.loads(json_str)
                else:
                    raise json.JSONDecodeError("No JSON found", content, 0)
                    
            except json.JSONDecodeError:
                parsed = {
                    "l1": "Thinking...", 
                    "l2": "Wait a sec >_<", 
                    "chat": "ìŒ... ë¬´ìŠ¨ ë§ì¸ì§€ ê³°ê³°ì´ ìƒê° ì¤‘ì´ì—ìš”.", 
                    "mood": "neutral", 
                    "act": "scan"
                }

            l1 = str(parsed.get("l1", "")).strip()
            l2 = str(parsed.get("l2", "")).strip()
            mood = str(parsed.get("mood", "neutral"))
            act = str(parsed.get("act", "none"))
            chat = str(parsed.get("chat", ""))

            # ë¹ˆ ì‘ë‹µ ë°©ì–´
            if not l1 and not chat:
                l1 = "Unknown..."
                l2 = "?_?"
                chat = "ë‹¤ì‹œ í•œ ë²ˆ ë§ì”€í•´ ì£¼ì‹œê² ì–´ìš”?"
                act = "scan"

            # 1. ë¶„ë…¸ í•„í„° (ê°•ì œ)
            if is_furious:
                mood = "angry"
                act = "shake"
                l1 = "-^-"
                l2 = "Sorry..."
                if "ì£„ì†¡" not in chat and "ë¯¸ì•ˆ" not in chat:
                    chat = "íˆìµ! ì œê°€ ì˜ëª»í–ˆì–´ìš”... ìš©ì„œí•´ì£¼ì„¸ìš” ã… ã… "

            # 2. ê°ì •-í–‰ë™ ë³´ì •
            if mood == "sad" and act == "nod": act = "none"
            elif mood == "angry" and act == "nod": act = "shake"

            # 3. LCD ì •ì œ ë° ì´ëª¨í‹°ì½˜ ê°•í™”
            def clean_and_enrich_lcd(text, mood, is_line2=False):
                # ASCIIë§Œ ë‚¨ê¸°ê¸°
                cleaned = "".join([c for c in text if ord(c) < 128]).strip()
                
                # ë‚´ìš©ì´ ì—†ìœ¼ë©´ ê¸°ë³¸ê°’
                if not cleaned:
                    if is_line2: cleaned = "..." 
                    else: cleaned = "Artie Robot"

                # ë„ˆë¬´ ê¸¸ë©´ ìë¥´ê¸° (ì´ëª¨í‹°ì½˜ ê³µê°„ í™•ë³´ë¥¼ ìœ„í•´ 13ì ì •ë„)
                if len(cleaned) > 13:
                    cleaned = cleaned[:13]
                
                # ëœë¤ ì´ëª¨í‹°ì½˜ ì¶”ê°€ (ì´ë¯¸ íŠ¹ìˆ˜ë¬¸ìê°€ ë§ì§€ ì•Šë‹¤ë©´)
                # ë¼ì¸ 2ì— ë” ì ê·¹ì ìœ¼ë¡œ ë„£ìŒ
                chance = 0.7 if is_line2 else 0.3
                if len(cleaned) < 13 and random.random() < chance:
                    if not re.search(r'[\^<>]', cleaned): 
                        emo_list = EMOTICONS.get(mood, EMOTICONS["neutral"])
                        emoji = random.choice(emo_list)
                        cleaned = f"{cleaned} {emoji}"
                
                return cleaned

            l1 = clean_and_enrich_lcd(l1, mood, is_line2=False)
            l2 = clean_and_enrich_lcd(l2, mood, is_line2=True)

            cmd = RobotCommand(
                seq=self.seq,
                l1=l1[:16],
                l2=l2[:16],
                chat_response=chat,
                mood=mood,
                act=act
            )
            self.result_signal.emit(cmd)

        except requests.exceptions.RequestException as e:
            # "[CRITICAL]" íƒœê·¸ë¥¼ ë¶™ì—¬ì„œ GUIì— ë³´ëƒ„
            self.error_signal.emit(f"[CRITICAL] API Connection Failed: {e}")

        except Exception as e:
            # ì¼ë°˜ì ì¸ íŒŒì´ì¬ ì—ëŸ¬
            self.error_signal.emit(f"LLM Error: {str(e)}")

# =========================
# GUI
# =========================
try:
    from_class = uic.loadUiType("LLM_With_Arduino.ui")[0]
except FileNotFoundError:
    from_class = QMainWindow

class LlmGui(QMainWindow, from_class):
    def __init__(self):
        super().__init__()
        try: self.setupUi(self)
        except AttributeError: self._init_ui_manually()

        self.setWindowTitle("Arduino Pet Robot - Artie (Gemma Mode)")
        self.hw = SmartController(SERIAL_PORT, BAUD_RATE)
        self.hw.sensor_received.connect(self.update_sensor_ui)
        
        self.inputButton.clicked.connect(self.process_input)
        self.inputLine.returnPressed.connect(self.process_input)
        
        self.is_closing_sequence = False
        
        self.last_light_status = "Normal"
        self.last_auto_trigger_time = 0
        self.prev_btn_state = 0
        
        self.append_system_msg("ì‹œìŠ¤í…œ ì‹œì‘. Gemma ë‘ë‡Œ íƒ‘ì¬ ì™„ë£Œ.")

    def _init_ui_manually(self):
        self.resize(600, 600)
        central = QWidget()
        self.setCentralWidget(central)
        layout = QVBoxLayout(central)

        self.lcdDisplay = QLabel("Initializing...", self)
        self.lcdDisplay.setStyleSheet("background-color: #99FF66; border: 2px solid #555; font-family: monospace; font-size: 20px; padding: 10px;")
        self.lcdDisplay.setAlignment(Qt.AlignmentFlag.AlignCenter)
        layout.addWidget(self.lcdDisplay)

        self.sensorLabel = QLabel("Waiting Sensor...", self)
        self.sensorLabel.setStyleSheet("font-weight: bold; font-size: 14px; color: #333;")
        layout.addWidget(self.sensorLabel)

        self.responseBox = QTextBrowser(self)
        layout.addWidget(self.responseBox)

        inp_layout = QHBoxLayout()
        self.inputLine = QLineEdit(self)
        self.inputButton = QPushButton("Send", self)
        inp_layout.addWidget(self.inputLine)
        inp_layout.addWidget(self.inputButton)
        layout.addLayout(inp_layout)

    def update_sensor_ui(self, data: SensorData):
        status_text = f"Temp: {data.temp}Â°C  |  Humid: {data.humid}%  |  Light: {data.light} ({data.light_status})"
        self.sensorLabel.setText(status_text)
        
        # 1. ì¡°ë„ ë³€í™” ìë™ ê°ì§€
        current_light = data.light_status
        now = time.time()
        
        if (current_light != self.last_light_status) and \
           (now - self.last_auto_trigger_time > 5.0) and \
           (not self.is_closing_sequence):
            
            if current_light in ["Dark", "Bright"]:
                print(f"âš¡ [Auto-Trigger] Light changed to {current_light}")
                self.last_auto_trigger_time = now
                self.trigger_auto_reaction(current_light)
            
            self.last_light_status = current_light
            
        # 2. ë²„íŠ¼ ì…ë ¥ ì²˜ë¦¬
        if data.btn == 1 and self.prev_btn_state == 0 and not self.is_closing_sequence:
            random_msg = random.choice(BUTTON_MESSAGES)
            self.lcdDisplay.setText(f"Btn: {random_msg[:11]}..")
            self.inputLine.setText(random_msg)
            self.process_input()
        
        self.prev_btn_state = data.btn

    def trigger_auto_reaction(self, light_status):
        msg = f"[í™˜ê²½ ë³€í™” ê°ì§€] ì¡°ëª… ìƒíƒœê°€ '{light_status}'ë¡œ ë³€ê²½ë˜ì—ˆìŠµë‹ˆë‹¤."
        self.append_system_msg(msg)
        
        query_text = f"(System Alert: The light level just changed to {light_status}. React to this sudden change immediately!)"
        
        self.lcdDisplay.setText("Sensing...")
        
        current_sensors = self.hw.get_sensor_data()
        self.worker = LLMWorker(query_text, self.hw.get_next_seq(), current_sensors)
        self.worker.result_signal.connect(self.on_llm_finish)
        self.worker.error_signal.connect(self.on_llm_error)
        self.worker.start()

    def process_input(self):
        if self.is_closing_sequence: return 

        raw_text = self.inputLine.text().strip()
        if not raw_text: return

        closing_keywords = ['quit', 'exit', 'ì¢…ë£Œ']
        is_shutdown_cmd = raw_text.lower() in closing_keywords
        
        query_text = raw_text
        
        if is_shutdown_cmd:
            print("ğŸ›‘ í”„ë¡œê·¸ë¨ ì¢…ë£Œ ìš”ì²­ë¨. ì‘ë³„ ì¸ì‚¬ë¥¼ ì‹œì‘í•©ë‹ˆë‹¤.")
            self.is_closing_sequence = True
            query_text = "System Termination"
            self.inputLine.setEnabled(False)
            self.inputButton.setEnabled(False)
        
        self.append_chat_msg("Me", raw_text, True)
        print(f"ğŸ—£ï¸ [Me] {raw_text}")
        
        self.inputLine.clear()
        if not is_shutdown_cmd:
            self.inputLine.setEnabled(False)
            
        if "Btn:" not in self.lcdDisplay.text():
             self.lcdDisplay.setText("Thinking...")
        
        current_sensors = self.hw.get_sensor_data()
        self.worker = LLMWorker(query_text, self.hw.get_next_seq(), current_sensors)
        self.worker.result_signal.connect(self.on_llm_finish)
        self.worker.error_signal.connect(self.on_llm_error)
        self.worker.start()

    def on_llm_finish(self, cmd: RobotCommand):
        self.hw.send_command(cmd)
        
        self.lcdDisplay.setText(f"{cmd.l1}\n{cmd.l2}")
        self.append_chat_msg("Artie", cmd.chat_response, False)
        
        print(f"ğŸ¤– [Artie] {cmd.chat_response}")
        print(f"   â†ª LCD: [{cmd.l1}] / [{cmd.l2}] | Mood: {cmd.mood} | Act: {cmd.act}")
        
        if self.is_closing_sequence:
            self.append_system_msg("3ì´ˆ í›„ ì‹œìŠ¤í…œì´ ì¢…ë£Œë©ë‹ˆë‹¤...")
            QTimer.singleShot(3000, self.close)
        else:
            self.inputLine.setEnabled(True)
            self.inputLine.setFocus()

    def on_llm_error(self, err: str):
        # ì‹œìŠ¤í…œ ë©”ì‹œì§€ ì¶œë ¥
        self.append_system_msg(f"Error: {err}")
        print(f"âš ï¸ [Error] {err}")
        
        # [ë³€ê²½] ì¹˜ëª…ì ì¸ API ì˜¤ë¥˜(400, ì—°ê²°ì‹¤íŒ¨ ë“±) ê°ì§€ ì‹œ ì¢…ë£Œ ì²˜ë¦¬
        if "[CRITICAL]" in err or "400" in err:
            self.append_system_msg("â›” ì¹˜ëª…ì ì¸ ì„œë²„ ì˜¤ë¥˜ê°€ ë°œìƒí–ˆìŠµë‹ˆë‹¤.")
            self.append_system_msg("3ì´ˆ í›„ í”„ë¡œê·¸ë¨ì„ ìë™ ì¢…ë£Œí•©ë‹ˆë‹¤...")
            
            # ì…ë ¥ ì°¨ë‹¨
            self.inputLine.setEnabled(False)
            self.inputButton.setEnabled(False)
            self.is_closing_sequence = True
            
            # 3ì´ˆ í›„ ì¢…ë£Œ (ì‚¬ìš©ìê°€ ì—ëŸ¬ ë©”ì‹œì§€ë¥¼ ì½ì„ ì‹œê°„ í™•ë³´)
            QTimer.singleShot(3000, self.close)
            return

        # ì¼ë°˜ ì—ëŸ¬ì¸ ê²½ìš° ê¸°ì¡´ ë¡œì§ ìˆ˜í–‰
        if self.is_closing_sequence:
            QTimer.singleShot(2000, self.close)
        else:
            self.inputLine.setEnabled(True)

    def append_chat_msg(self, sender, text, is_user):
        align = "right" if is_user else "left"
        color = "#e6f2ff" if is_user else "#ffffff"
        self.responseBox.append(f"<div style='text-align:{align};'><span style='background:{color};padding:5px;'><b>{sender}</b>: {text}</span></div>")

    def append_system_msg(self, text):
        self.responseBox.append(f"<div style='text-align:center;color:gray;font-size:small;'>{text}</div>")

    def closeEvent(self, event):
        self.hw.close()
        print("âœ… ì‹œìŠ¤í…œì´ ì•ˆì „í•˜ê²Œ ì¢…ë£Œë˜ì—ˆìŠµë‹ˆë‹¤.")
        event.accept()

def main():
    app = QApplication(sys.argv)
    win = LlmGui()
    win.show()
    sys.exit(app.exec())

if __name__ == "__main__":
    main()